src/main/java/com/getindata/connectors/http/internal/security/SecurityContext.java

```diff
@@ -0,0 +1,184 @@
+            CertificateFactory certificateFactory = CertificateFactory.getInstance("X.509");
+            Certificate certificate = certificateFactory.generateCertificate(certInputStream);
+            this.keystore.setCertificateEntry(UUID.randomUUID().toString(), certificate);
+            log.info("Certificated added to keyStore ass trusted - " + certPath);
+        } catch (Exception e) {
+            throw new RuntimeException(
+                "Unable to add certificate as trusted to Http Connector security context - "
+                    + certPath,
+                e
+            );
+        }
+    }
+
+    /**
+     * Add certificate and private key that should be used by anny Http Connector instance that uses
+     * this {@link SSLContext} instance. Certificate and key are added only to this Context's
+     * {@link KeyStore} and not for entire JVM.
+     * @param publicKeyPath path to public key/certificate used for mTLS.
+     * @param privateKeyPath path to private key used for mTLS.
+     */
+    public void addMTlsCerts(String publicKeyPath, String privateKeyPath) {
+
+        try {
+            byte[] publicData = Files.readAllBytes(Path.of(publicKeyPath));
+            byte[] privateData = Files.readAllBytes(Path.of(privateKeyPath));
+            byte[] decodedPrivateData = decodePrivateData(privateKeyPath, privateData);
+
+            CertificateFactory certificateFactory = CertificateFactory.getInstance("X.509");
+            Collection<? extends Certificate> chain = certificateFactory.generateCertificates(
+                new ByteArrayInputStream(publicData));
+
+            Key key = KeyFactory.getInstance("RSA")
+                .generatePrivate(new PKCS8EncodedKeySpec(decodedPrivateData));
+
+            this.keystore.setKeyEntry(
+                UUID.randomUUID().toString(),
+                key,
+                this.storePasswordCharArr,
+                chain.toArray(new Certificate[0])
+            );
+        } catch (Exception e) {
+            throw new RuntimeException(
+                "Unable to add client private key/public certificate to Http Connector KeyStore. "
+                    + String.join(",", privateKeyPath, publicKeyPath),
+                e
+            );
+        }
+    }
+
+    /**
+     * Reads private key data. Key can be in PEM and DER coding and in PKCS8 format.
+     * @param privateKeyPath path to private key.
+     * @param privateData read bytes from private key,
+     * @return decoded key data.
+     */
+    private byte[] decodePrivateData(String privateKeyPath, byte[] privateData) {
+
+        // private key must be in PKCS8 format, pem or der.
+        // openssl pkcs8 -topk8 -inform PEM -outform PEM -in client.pem
+        // -out clientPrivateKey.pem -nocrypt
+        if (privateKeyPath.endsWith(".pem")) {
+            String privateString = new String(privateData, Charset.defaultCharset())
+                .replace("-----BEGIN PRIVATE KEY-----", "")
```

Comment:
According to [Wikipedia page on PKCS8](https://en.wikipedia.org/wiki/PKCS_8) and the constants at the beginning of [<openssl/pem.h>](https://git.openssl.org/?p=openssl.git;a=blob;f=include/openssl/pem.h;hb=HEAD), PKCS8 can start with not only `BEGIN PRIVATE KEY` but also `BEGIN ENCRYPTED PRIVATE KEY`. Also, I am not so sure about matching of exactly five hyphens — quick search in the Google doesn't show usage of different number of hyphens, but it gives me a bit of shaky hands 😬
And why `END PRIVATE KEY` replacement happens _after_ replacing newlines? I suppose it could be a bit safer first to replace both begin and end markers, and then go with the newlines in base64 inners.

Reply:
Self-update: according to [RFC 7468](https://www.rfc-editor.org/rfc/rfc7468#section-2), _there are exactly five hyphen-minus characters_ 🙂